### PR TITLE
[3226][ADD] purchase_stock_bl_date

### PR DESCRIPTION
**Pull request by AungKoKoLin1997**
[3226](https://www.quartile.co/web#id=3226&action=771&model=project.task&view_type=form&menu_id=505)
depends on : https://github.com/qrtl/hls-oca/pull/31

### CODE REVIEW

File: purchase_order_bl_date/models/stock_move.py [hunk 1]
```diff
@@ -0,0 +1,28 @@
...
+
+from odoo import models, api, fields
+import logging
+
+
+class StockMove(models.Model):
+    _inherit = 'stock.move'
+
+    @api.multi
+    def _get_price_unit(self):
```
Comment (kanda999) [inline purchase_order_bl_date/models/stock_move.py]: Is it possible to override `_get_conversion_rate` and pass `bl_date`? 
we would like to avoid a complete override.

Another my concern is that when you `button_confirm` in purhase_order with `bl_date`,  price_unit of the stock.move record that is created is not affected by the bl_date.
Because if I am not mistaken, the method `_get_stock_move_price_unit` is used when the stock.move record is first created and this `_get_price_unit` is not used. 

Can you check that is it a problem that stock.move is calculated and created at today's rate even if the `bl_date` is set?

File: purchase_order_bl_date/views/stock_picking_views.xml [hunk 2]
```diff
@@ -0,0 +1,12 @@
...
+        <field name="name">stock.picking.form.inherit</field>
+        <field name="model">stock.picking</field>
+        <field name="inherit_id" ref="stock.view_picking_form"/>
+        <field name="arch" type="xml">
+            <field name="group_id" position="after">
+                <field name="bl_date"/>
+            </field>
+        </field>
+    </record>
+</odoo>
```
Comment (kanda999) [inline purchase_order_bl_date/views/stock_picking_views.xml]: Please run pre-commit
```suggestion
</odoo>

```

**Review by kanda999 — CHANGES_REQUESTED** [review]

File: purchase_order_bl_date/models/stock_move.py (hunk 1 above)
Comment (AungKoKoLin1997) [reply purchase_order_bl_date/models/stock_move.py]: @kanda999 
The `_get_conversion_rate` is not called from `stock` module when picking is created from purchase or when they create stock valuation. So, I have no idea how to `_get_conversion_rate` and pass `bl_date`. I also wanted to avoid complete override `_get_price_unit` but this function only returns the converted price_unit and we can't convert again. And I see odoo override like this https://github.com/odoo/odoo/blob/0e56e4a7bd6de42d729441a53995ddd459d5e633/addons/purchase_stock/models/stock.py#L37.

I also knew `_get_stock_move_price_unit` is called when we click `button_confirm` in `purchase_order`. But when we calculate stock valuation and journal entry, it works with _get_price_unit(). But I agree we should use _get_stock_move_price_unit for design consistent.

File: purchase_order_bl_date/models/account_invoice.py [hunk 3]
```diff
@@ -0,0 +1,15 @@
...
+
+class AccountInvoice(models.Model):
+    _inherit = "account.invoice"
+
+    bl_date = fields.Date()
+
+    def _get_currency_rate_date(self):   
+        if self.bl_date:
+            return self.bl_date
+        return super()._get_currency_rate_date()
```
Comment (yostashiro) [inline purchase_order_bl_date/models/account_invoice.py]: Not perfectly sure but I don't think we need this. Invoice date = B/L date per my understanding.

File: purchase_order_bl_date/models/stock_move.py (hunk 1 above)
Comment (yostashiro) [reply purchase_order_bl_date/models/stock_move.py]: Can we do like this:

```python
    @api.multi
    def _get_price_unit(self):
        self.ensure_one()
        bl_date = self.picking_id.bl_date
        if bl_date:
            self = self.with_context(bl_date=bl_date)
        return super()._get_price_unit()
```

And then (in `res.currency`):

```python
    def _convert(self, from_amount, to_currency, company, date, round=True):
        bl_date = self._context.get("bl_date")
        date = bl_date or date
        return _convert(from_amount=fro_amount, to_currency=to_currency, company=company, date=date, round=round)
```

**Review by yostashiro — CHANGES_REQUESTED** [review]

**Comment by yostashiro** [comment]
Can we rename the module to `purchase_stock_bl_date`?

File: purchase_stock_bl_date/__manifest__.py [hunk 4]
```diff
@@ -0,0 +1,14 @@
+# Copyright 2023 Quartile Limited
+# License LGPL-3.0 or later (http://www.gnu.org/licenses/lgpl).
+{
+    "name": "Purchase Stock Bl Date",
+    "summary": "",
+    "version": "12.0.1.0.0",
+    "category": "Purchase",
+    "website": "https://www.quartile.co",
+    "author": "Quartile Limited",
+    "license": "LGPL-3",
```
Comment (yostashiro) [inline purchase_stock_bl_date/__manifest__.py]: Any reason to make it LGPL?

File: purchase_stock_bl_date/models/purchase.py [hunk 5]
```diff
@@ -0,0 +1,39 @@
...
+    bl_date = fields.Date()
+
+    @api.multi
+    def action_view_invoice(self):
+        res = super(PurchaseOrder, self).action_view_invoice()
+        if res.get("context"):
+            res["context"].update(
+                default_date_invoice=self.bl_date,
+            )
+        return res
```
Comment (yostashiro) [inline purchase_stock_bl_date/models/purchase.py]: Can we remove this for now?  From my point of view, the user should just input in the invoice date field the date on the invoice from the supplier, and there is no difference whether there is B/L date or not (not perfectly sure, but that is my assumption).

```suggestion
```

File: purchase_stock_bl_date/models/purchase.py [hunk 6]
```diff
@@ -0,0 +1,39 @@
...
+class PurchaseOrderLine(models.Model):
+    _inherit = "purchase.order.line"
+
+    @api.multi
+    def _get_stock_move_price_unit(self):
+        self.ensure_one()
+        bl_date = self.order_id.bl_date
+        if bl_date:
+            self = self.with_context(bl_date=bl_date)
+        return super()._get_stock_move_price_unit()
```
Comment (yostashiro) [inline purchase_stock_bl_date/models/purchase.py]: IMO, we should remove this (unless I'm overlooking something).  There is be no difference whether B/L date is available or not - when an incoming move is created from a PO, we just follow the standard logic.  What matters is the correct application of the unit_price where it affects the stock valuation or journal entries.  Everything else does not matter.

```suggestion
```

File: purchase_stock_bl_date/models/stock_move.py [hunk 7]
```diff
@@ -0,0 +1,27 @@
...
+    def _create_account_move_line(
+        self, credit_account_id, debit_account_id, journal_id
+    ):
+        self.ensure_one()
+        bl_date = self.picking_id.bl_date
+        if bl_date:
+            self = self.with_context(force_period_date=bl_date)
+        return super()._create_account_move_line(
+            credit_account_id, debit_account_id, journal_id
+        )
```
Comment (yostashiro) [inline purchase_stock_bl_date/models/stock_move.py]: This is out of scope of the enhancement this time.  Although we might want to seek a way to provide an option for users to force a date in pickings, the solution should be implemented in a comprehensive manner, not only for the case where B/L date is available.

```suggestion
```

File: purchase_stock_bl_date/__manifest__.py [hunk 8]
```diff
@@ -0,0 +1,14 @@
+# Copyright 2023 Quartile Limited
+# License LGPL-3.0 or later (http://www.gnu.org/licenses/lgpl).
+{
+    "name": "Purchase Stock Bl Date",
+    "summary": "",
```
Comment (yostashiro) [inline purchase_stock_bl_date/__manifest__.py]:
```suggestion
```

File: purchase_stock_bl_date/readme/DESCRIPTION.rst [hunk 9]
```diff
@@ -0,0 +1,5 @@
+This modules does the following:
+
+* Add bl_date in purchase order.
+* Pass bl_date to bill and recipts
+* Calulate journal entry with bl_date rate
```
Comment (yostashiro) [inline purchase_stock_bl_date/readme/DESCRIPTION.rst]:
```suggestion
```

Please improve the description with business background.

**Review by yostashiro — CHANGES_REQUESTED** [review]

File: purchase_stock_bl_date/readme/DESCRIPTION.rst [hunk 10]
```diff
@@ -1,5 +1,12 @@
...
-* Calulate journal entry with bl_date rate
+* Pass bl_date recipts
+
+Background
+
+In standard Odoo, the system calculates the exchange rate of journal entries based on the current date
+when generating stock valuations. This can cause inconvenience if users need to compute the rate of
+journal entries based on a custom date, especially when the currency used in the record differs from
+the company's currency. This module addresses this issue by enabling the computation of the exchange rate
+based on a user-defined date.
```
Comment (rinaldifirdaus) [inline purchase_stock_bl_date/readme/DESCRIPTION.rst]: @AungKoKoLin1997 , Kindly check following description:

This modules does the following:

* Add bl_date in purchase order.
* Pass bl_date to receipts

Background

In an operation where the purchase order is created at the same time as the receipt of goods, which is done due to frequent discrepancies between the received quantity of goods and the quantity at the time of receipt, there is a condition where the currency rate in effect at the time of creating the purchase order differs from the currency rate used when the B/L (Bill of Lading) is issued.
This becomes a problem when the user wants the price calculated in the journal entry to use the currency rate used on the B/L issuance date. 

This module is designed to address this issue by calculating the price in the journal entry based on the currency rate in effect on the specified "B/L Date" in purchase order.

Comment (kanda999) [reply purchase_stock_bl_date/readme/DESCRIPTION.rst]: >  done due to frequent discrepancies between the received quantity of goods and the quantity at the time of receipt

Not correct.
When purchasing product, you want to apply the rate of that day (BL date) because the product belong to the company from the moment the product come in ship.
Customers want to apply the rate at the time the product are placed on the ship.

Comment (AungKoKoLin1997) [reply purchase_stock_bl_date/readme/DESCRIPTION.rst]: @kanda999  @rinaldifirdaus  Is the following correct?

Background

During the procurement process, where the purchase order and receipt of goods occur simultaneously, discrepancies can frequently arise between the received quantity of goods and the quantity recorded at the time of receipt. This can also result in a difference between the currency rate in effect at the time of creating the purchase order and the currency rate in effect when the B/L (Bill of Lading) is issued.

This issue becomes critical when the user prefers to have the price calculated in the journal entry using the currency rate prevalent on the B/L issuance date rather than current date. This is because from the moment the products are loaded onto the ship (B/L date), they belong to the company. Therefore, customers want to apply the currency rate of the day when the products are placed on the ship.

This module addresses this requirement by calculating the price in the journal entry based on the currency rate in effect on the specified "B/L Date" in the purchase order. This ensures a more accurate reflection of costs according to when the company effectively takes ownership of the goods.

File: purchase_stock_bl_date/readme/DESCRIPTION.rst [hunk 11]
```diff
@@ -0,0 +1,15 @@
...
+Background
+
+In an operation where the purchase order is created at the same time as the receipt of goods, which is done due to
+frequent discrepancies between the received quantity of goods and the quantity at the time of receipt, there is a
+condition where the currency rate in effect at the time of creating the purchase order differs from the currency rate
+used when the B/L (Bill of Lading) is issued. This becomes a problem when the user wants the price calculated in the
+journal entry to use the currency rate used on the B/L issuance date.
+
+This module is designed to address this issue by calculating the price in the journal entry based on the currency rate
+in effect on the specified "B/L Date" in purchase order.
```
Comment (yostashiro) [inline purchase_stock_bl_date/readme/DESCRIPTION.rst]: Here is my version:

```suggestion
This module adds B/L Date fields to the purchase order and the picking.  B/L Date assigned in the purchase receipt is used for capturing the currency rate for the unit price calculation, which is the basis of inventory valuation.

Note that this module only adjusts the logic where currency rate is captured.  Adjusting the date of stock moves or journal entries is outside the scope of this module.

Background:
~~~~~~~~~~~

Depending on the incoterms of the purchase transaction in international trades, the ownership transfer of the purchased products takes place before they arrive at the demestic warehouse (i.e. when products are loaded in cargo in the origin country), while the receipt operation in Odoo, for practical reasons, have to be done when the products are received at the warehouse.  This results in an inaccurate valuation of the inventory, which should be avoided.
```

Does this make sense?

Please also add USAGE.rst.

**Review by yostashiro — CHANGES_REQUESTED** [review]
@rinaldifirdaus Please check if the recovery can be done under the current design in case user makes mistake with the operation (e.g. forget to input B/L date).

If the recovery operation is tricky (which I suppose it is), please suggest and adjust the desgin with Lin/Kanda-san.

File: purchase_stock_bl_date/readme/USAGE.rst [hunk 12]
```diff
@@ -0,0 +1,3 @@
+* Go to the purchase
+* Create new quotation or edit existing quotation
+* Add bl_date field under Other Information tab.
```
Comment (yostashiro) [inline purchase_stock_bl_date/readme/USAGE.rst]: This doesn't provide sufficient information.

@rinaldifirdaus Please help improve this.

Comment (rinaldifirdaus) [reply purchase_stock_bl_date/readme/USAGE.rst]: Adjust the usage.rst on the new commit.
please kindly your review.

File: purchase_stock_bl_date/readme/USAGE.rst [hunk 13]
```diff
@@ -0,0 +1,15 @@
...
+
+2. Validate the related stock picking
+
+* On purchase order, click "Receive product" button
+* On picking, make sure the bl_date has been reflected on the Other Information tab, then validate the picking
+
+3. Check computed amount on generated Journal Entry
+
+* Go to Invoicing> Accounting> Journal Entry
+* Open the stock valuation entry, make sure the amount calculated with the rate that applied on the specified bl_date
```
Comment (yostashiro) [inline purchase_stock_bl_date/readme/USAGE.rst]: Here is my version:

```suggestion
* Set B/L Date in purchase order as appropriate. When the purchase order is confirmed and the receipt picking is generated, the set date is passed over to the corresponding field of the picking.
* Adjust B/L Date of the receipt picking as necessary before validating the receipt.
```

Who is our audience? It isn't end users. Instead, it's ourselves, individuals with business experience and the capability to read the code. Let's be clear on this: We are writing for our future selves.

In this context, what we aim to explain here is the design concept rather than providing step-by-step instructions for end users. I understand that the README files of some OCA modules are written in this style. However, I've noticed that these documents often get neglected during design changes because verbosity increases the maintenance cost.

File: purchase_stock_bl_date/readme/DESCRIPTION.rst (hunk 11 above)
Comment (rinaldifirdaus) [reply purchase_stock_bl_date/readme/DESCRIPTION.rst]: >Please check if the recovery can be done under the current design in case user makes mistake with the 
>operation (e.g. forget to input B/L date).
>If the recovery operation is tricky (which I suppose it is), please suggest and adjust the desgin with Lin/Kanda-san.

Hi @yostashiro san, as you mentioned that if user forgot to set the bl date, the recovery is tricky with current design.

@AungKoKoLin1997 
Im thinking following logic, is it possible to be implemented?
- set the bl_date as read only when picking state is done (currently bl date is editable even state is done)
- add new boolean field "force bl date" after bl_date, editable only if user click unlock button on picking
- if "force bl date" is True, bl_date is editable
- if bl_date updated on picking in Done state, and user click lock button, its trigger following process:
  1. revaluate the stock valuation with the updated bl_date
  2. reverse the current related svl journal entry, and generate new journal entry with the updated information. (or if possible to directly update the current related svl journal entry is better)

Kindly your advise

**Comment by AungKoKoLin1997** [comment]
This PR is ready for review.

File: purchase_stock_bl_date/models/stock_picking.py [hunk 14]
```diff
@@ -0,0 +1,26 @@
...
+                move.write(
+                    {
+                        "price_unit": move._get_price_unit(),
+                    }
+                )
+                if move.product_id.cost_method == "average":
+                    move.cost_price_avco_sync()
+                else:
+                    move._run_valuation()
+                move.update_amount_journal_entry()
```
Comment (yostashiro) [inline purchase_stock_bl_date/models/stock_picking.py:25]: Am I correct to understand that this simply overrides the value of the receipt stock move and its journal entry?  In that case, it adds the risk of inconsistent balances in case the inventory has already been shipped out (in FIFO context).  Is there a way to prevent that from happening (I suppose the same applies to AVCO - has this issue/concern been addressed somewhere)?

Comment (AungKoKoLin1997) [reply purchase_stock_bl_date/models/stock_picking.py:25]: Yes, you right. I have no idea how to prevent in case the inventory has already been shipped out at the moment.

Comment (AungKoKoLin1997) [reply purchase_stock_bl_date/models/stock_picking.py:25]: @kanda999 @rinaldifirdaus  Please let me know if you guys have the advice for this case. Current situation for recovery can recalculate the journal entry and stock valuation.

Comment (rinaldifirdaus) [reply purchase_stock_bl_date/models/stock_picking.py:25]: in manual operation, to do recovery of the inventory that has been shipped out, we can adjust the product's done qty on the delivery record to be 0 (same with the logic on receipt). when we set the delivered qty to 0, valuation of the product will be back to as it is on inventory valuation layer regardless of the current applied unit price.
then the flow would be:
1. manually set the delivered qty to 0 (valuation will be corrected with amount when it was delivered)
2. set the force bl date as true, update/revise the bl date 
3. manually set back the delivered qty to its original qty on the delivery record (valuation will be corrected with the new amount based on bl date correction)

i think we may need to ask the client to recover the delivery manually (without customization) then update the bl date as above flow. 
or maybe you have an idea to automate above flow @AungKoKoLin1997 ?
it might be tricky to automate set the delivered qty to 0 since on the delivery record may have various products (not only the inventory that we want to recover) on the SO line. but it might be ok to temporarily set all SO line qty to 0 and set back all the SO line qty as it was after update/revise the bill date (need to check).

Comment (AungKoKoLin1997) [reply purchase_stock_bl_date/models/stock_picking.py:25]: >  maybe you have an idea to automate above flow

It is very hard for me to implement this and this is very complex.

**Review by kanda999 — APPROVED** [review]
Functional review

File: purchase_stock_bl_date/views/purchase_order_views.xml [hunk 15]
```diff
@@ -0,0 +1,13 @@
+<?xml version="1.0" encoding="utf-8"?>
+<odoo>
+    <record id="view_order_form" model="ir.ui.view">
+        <field name="name">purchase.order.form</field>
+        <field name="model">purchase.order</field>
+        <field name="inherit_id" ref="purchase.purchase_order_form"/>
+        <field name="arch" type="xml">
+            <xpath expr="//form/sheet/group[1]/group[2]" position='inside'>
```
Comment (yostashiro) [inline purchase_stock_bl_date/views/purchase_order_views.xml]:
```suggestion
            <xpath expr="//field[@name='date_order']/.." position='inside'>
```
This should be safer and more explicit.

File: purchase_stock_bl_date/models/account_move_line.py [hunk 16]
```diff
@@ -0,0 +1,14 @@
...
+from odoo import api, models
+
+
+class AccountMoveLine(models.Model):
+    _inherit = "account.move.line"
+
+    @api.multi
+    def _update_check(self):
+        if self.env.context.get("skip_check_update"):
+            return True
```
Comment (yostashiro) [inline purchase_stock_bl_date/models/account_move_line.py:15]: Comment the rationale of this.

File: purchase_stock_bl_date/readme/USAGE.rst [hunk 17]
```diff
@@ -0,0 +1,11 @@
...
+  the receipt picking is generated, the set date is passed over to the corresponding field of the picking.
+* Adjust B/L Date of the receipt picking as necessary before validating the receipt.
+
+In case you forget to assign or need to change the B/L Date after the receipts have been
+validated, follow these steps:
+
+* Open the Receipts form and click on the "Unlock" button.
+* Enable the 'Force B/L Date' option and add or update the B/L Date.
+* After updating, click on the "Lock" button to update the inventory valuation and
+  the corresponding Stock Journal Entry.
```
Comment (yostashiro) [inline purchase_stock_bl_date/readme/USAGE.rst]: This concern https://github.com/qrtl/hls-custom/pull/209/files#r1263209617 must be addressed.

File: purchase_stock_bl_date/models/stock_picking.py [hunk 18]
```diff
@@ -0,0 +1,26 @@
...
+from odoo import fields, models
+
+
+class StockPicking(models.Model):
+    _inherit = "stock.picking"
+
+    bl_date = fields.Date(string="B/L Date")
+    force_bl_date = fields.Boolean(string="Force B/L Date")
+
+    def action_toggle_is_locked(self):
```
Comment (yostashiro) [inline purchase_stock_bl_date/models/stock_picking.py]: I am not comfortable with this method being the trigger to update the past transactions as the implication is not obvious to the user. How about adding a "Update Valuation" button below B/L Date field, which updates past transactions, and then triggers action_toggle_is_locked() to make itself invisible?

File: purchase_stock_bl_date/models/stock_move.py [hunk 19]
```diff
@@ -0,0 +1,30 @@
...
+                ("stock_move_id", "=", self.id),
+                ("state", "=", "posted"),
+                ("company_id", "=", self.env.user.company_id.id),
+            ]
+        )
+        bl_date = self.picking_id.bl_date
+        if bl_date:
+            move.with_context(
+                bl_date=bl_date, skip_check_update=True, check_move_validity=False
+            )._onchange_date()
```
Comment (yostashiro) [inline purchase_stock_bl_date/models/stock_move.py]:
```suggestion
        bl_date = self.picking_id.bl_date
        if not bl_date:
            return
        # There can be multiple account moves in case a correction entry has been generated.
        account_moves = self.env["account.move"].search(
            [
                ("stock_move_id", "=", self.id),
                ("state", "=", "posted"),
                ("company_id", "=", self.env.user.company_id.id),
            ]
        )
        for account_move in account_moves:
            account_move.with_context(
                bl_date=bl_date, skip_check_update=True, check_move_validity=False
            )._onchange_date()
```

**Review by yostashiro — CHANGES_REQUESTED** [review]

File: purchase_stock_bl_date/readme/USAGE.rst (hunk 17 above)
Comment (rinaldifirdaus) [reply purchase_stock_bl_date/readme/USAGE.rst]: @AungKoKoLin1997 , Regarding this concern, is there any way to do checking whether the stock of the product that received on the transfer already shipped out or not? i guess it will be hard..
but if there's any way to do that, i think we shall add that checking and block the B/L dates update if the stock has been shipped out with user error and give information to manually adjust the related delivery to be able to update the B/L date. kindly your advise.

File: purchase_stock_bl_date/models/stock_picking.py (hunk 18 above)
Comment (rinaldifirdaus) [reply purchase_stock_bl_date/models/stock_picking.py]: I agree with tashiro san suggestion. can you please help to follow up @AungKoKoLin1997 ?

**Comment by rinaldifirdaus** [comment]
I've added new commit to make the 'update valuation' button invisible when the force bl date is false to make the button is clickable when the b/l date value need to be updated.

**Review by yostashiro — APPROVED** [review]
Code review.

**Review by rinaldifirdaus — APPROVED** [review]
@kanda999 san, can we have this PR merged?
